text/0048-session-keys-runtime-api.md

```diff
@@ -0,0 +1,115 @@
+# RFC-0048: Generate ownership proof for `SessionKeys`
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 13 November 2023                                                                    |
+| **Description** | Change `SessionKeys` runtime api to support generating an ownership proof for the on chain registration. |
+| **Authors**     | Bastian Köcher                                                                    |
+
+## Summary
+
+This RFC proposes to changes the `SessionKeys::generate_session_keys` runtime api interface. This runtime api is used by validator operators to
```

Comment:
Because the node doesn't know what session keys the runtime is using.

Reply:
Instead of generating the keys in the runtime, which in fact ends up generating the keys via some host call would be nice to have instead the runtime returning a list of `(CryptoTypeId, KeyTypeId)`.
Then generate the keys directly using the keystore.

The advantage of this approach would be to deprecate the host calls to generate the keys from the runtime which (IMO) is a bit awkward.

Reply:
These host calls are also required for benchmarking to generate private keys. `CryptoTypeId` didn't exist at the time this runtime api was introduced. This is also really the topic of this RFC. 

In the end you also still need to call into the runtime to do this. In hindsight I would also design some of the host calls differently to make nodes for example "more generic". This can be done, but yeah isn't part of this RFC and this RFC will also not make any of that more complicated.